src/notation/view/notationviewinputcontroller.cpp

```diff
@@ -246,6 +252,79 @@ void NotationViewInputController::setViewMode(const ViewMode& viewMode)
     notation->setViewMode(viewMode);
 }
 
+constexpr qreal scrollStep = .8;
+constexpr qreal thinPadding = 10.0;
+constexpr qreal thickPadding = 25.0;
+
+void NotationViewInputController::moveScreen(int direction)
+{
+    auto notation = currentNotation();
+    if (!notation) {
+        return;
+    }
+    auto scale = m_view->currentScaling();
+    if (notation->viewMode() == ViewMode::LINE) {
+        m_view->moveCanvasHorizontal(m_view->width() * direction * scrollStep / scale);
+    } else {
+        m_view->moveCanvasVertical(m_view->height() * direction * scrollStep / scale);
+    }
+}
+
+void NotationViewInputController::movePage(int direction)
+{
+    auto notation = currentNotation();
+    if (!notation) {
+        return;
+    }
+    if (notation->viewMode() != ViewMode::PAGE) {
+        moveScreen(direction);
+        return;
+    }
+    Page* page = notation->elements()->msScore()->pages().back();
+    if (configuration()->canvasOrientation().val == mu::framework::Orientation::Vertical) {
+        qreal offset = std::min((page->height() + thickPadding) * direction, m_view->toLogical(QPoint()).y() + thickPadding);
+        m_view->moveCanvasVertical(offset);
+    } else {
+        qreal offset = std::min((page->width() + thickPadding) * direction, m_view->toLogical(QPoint()).x() + thickPadding);
+        m_view->moveCanvasHorizontal(offset);
+    }
+}
+
+void NotationViewInputController::nextScreen()
+{
+    moveScreen(-1);
+}
+
+void NotationViewInputController::previousScreen()
+{
+    moveScreen(1);
+}
+
+void NotationViewInputController::nextPage()
+{
+    movePage(-1);
+}
+
+void NotationViewInputController::previousPage()
+{
+    movePage(1);
+}
+
+void NotationViewInputController::startOfScore()
+{
+    auto offset = m_view->toLogical(QPoint());
+    m_view->moveCanvas(offset.x() + thickPadding, offset.y() + thickPadding);
+}
+
+void NotationViewInputController::endOfScore()
+{
+    auto offset = m_view->toLogical(QPoint());
+    auto rect = m_view->notationContentRect();
+    auto scale = m_view->currentScaling();
+    m_view->moveCanvas(offset.x() + m_view->width() / scale - thickPadding - rect.width(),
+                       offset.y() + m_view->height() / scale - thickPadding - rect.height());
+}
```

Comment:
Actually this doesn't work properly for the case where the last page isn't "full" - it always scrolls to the bottom of the final page.  Needs a bit more thought.